### ISSUE TITLE
How to generate subtasks for pi0.5?

### ISSUE THREAD

**Issue by chinaydp** [Physical-Intelligence/openpi#679]
both #647 and #664 have mentioned the issue of subtask prediction，I attempted to modify the sample_actions method in the pi0 class by replacing:
` _, kv_cache = self.PaliGemma.llm([prefix_tokens, None], mask=prefix_attn_mask, positions=positions)` with:

```
(prefix_out, suffix_out), kv_cache = self.PaliGemma.llm([prefix_tokens, None], mask=prefix_attn_mask, positions=positions) 
logits = self.PaliGemma.llm(prefix_out, method='decode') 
token_ids = jnp.argmax(logits, axis=-1) 
token_list = token_ids[0].tolist() 
decoded_text = self._tokenizer.decode(token_list) 
```
to generate subtasks. However, I got a bunch of garbled words. What should I do to generate subtasks?

**Comment by Marxlp** [comment]
Same question, is the released model with implicit HL?

<img width="567" height="371" alt="Image" src="https://github.com/user-attachments/assets/92bdf29a-a922-4f53-8962-b81d22292901" />
@kvablack

**Comment by tahsinkose** [comment]
Similar to @chinaydp I replaced the [one-liner](https://github.com/Physical-Intelligence/openpi/blob/0d3eb2db836e443a6ca9debb242d0aabe4628f1f/src/openpi/models/pi0.py#L237) with:
```
(prefix_out, _), kv_cache = self.PaliGemma.llm([prefix_tokens, None], mask=prefix_attn_mask, positions=positions)
# VLM introspection: decode what the model is "thinking" in natural language
if prefix_out is not None:
    logits = self.PaliGemma.llm(prefix_out, method="decode")
    token_ids = jnp.argmax(logits, axis=-1)
    # Use debug callback to decode text outside JIT context
    def decode_and_print(token_ids_array):
        try:
            token_list = token_ids_array.tolist()
            decoded_text = self.tokenizer._tokenizer.decode(token_list)
            print(f"VLM decoded text (inference): {decoded_text}")
        except Exception as e:
            print(f"VLM decoding error (inference): {e}")
    jax.debug.callback(decode_and_print, token_ids[0])
```

I get a garbled output as follows:
```
VLM decoded text (inference): ⊏넙鬈∐Ⅽ♼㊄ϫ≎넙⊻⊏Ἅ⊏ up㎍넙 up∐Ѐ㊄Ⓧϫ∐ἍⅭ솝㉗㎍ἍSub upSubSub㉗㉗ꪑ⊻㉗ꪑЀӔЀЀ upἍ넙 up㉗Sub㉗㉗㉗Ѐ㉗∐ЀЀЀ⊏㉗⍤SubSub㉗Subኹ㉗Ѐ∐㉗ϫ∐∐∐鍩㉗Yes넙㉗㉗Ѐ㉗㉗ⅭⅭ㉗舠Ἅ∐Ⓧ⍀㉗ӒSub㉗㉗㉗㉗㉗Ⅽ舠뻑舠Ἅ⊻Ἅ舠㉗㉗Sub㉗㉗Ѐ㉗Ἅ立ϫ立蛲짭짭立蛲舠쫑㉗㉗Ἅ넙㉗㉗≎ⅭӒ蛲ӒἍ蛲鍩ӒἍSub㉗㉗Ѐ㉗Ἅଝ쐈蛲蛲냑냑蛲蛲Ӓ㉗SubSubኹἍ㉗㉗ଝ嫫ଝ蛲Ӓ立蛲ϫ㉗Ἅ㉗鬈㉗㉗㉗㉗Ӓ⊻Ἅ蛲立ЀἍӒ㉗㉗㉗⊏㉗㉗㉗鬈蛲⊻ἍἍ立Ἅ蛲㉗㉗Sub⍤ up㉗㉗㉗ኹἍᎱἍἍ立ἍἍꪑ㉗∐⊻∐⊻∐⊻鍩Ѐ立蛲蛲蛲舠蛲⊻立⊏넙Ἅ∐넙∐늠NoNo蛲蛲ЀNo∐堟∐넙∐ the鎴ꯎ∐ꯎ∐∐Yes the鎴 the∐YesYes the∐ theNoYesYesNo up鎴 up the∐No㎍ upNo the∐∐㎍∐ theYes鎴Noadjust∐∐鎴鎴⊏ theadjust the鎴∐鎴 the the the up㎍㎍NoNo the the the鎴 up up∐ up the the鎴adjustadjustadjust the the theadjust∐鎴∐∐鎴∐∐⊏⊏∐∐∐∐∐鎴∐∐ up넙Ⓧ up立 the⊏ӝ∐⊏Ѐӝ蘗Yes畎Yes⊻囷ⓙ푔ꯎꯎ upଝꪑ⢺ꯎ⤙ⓙ∐立堟SubЀ the inSub upSubNoSub up立∐㎍立ʶ up theSubSub upSubꪑSub뗄㎍ꪑᵊSub퐹∐∐SubSubSub㎍∐Subꪑ⊏立뗄⊏넙 up퐹∐⊏Sub 堟㎍∐㎍∐⊏∐໑鐗넙ᵊᵊꪑፁSubЀ⊏㎍🅺鐗Ҥ짹鐗⊏Sub鐽ᵊꪑ⊏SubSub⊏⊏堟Ѐⴳﮗ∐立∐鐗ЀꪑNoSubSubЀ⊏ꪑЀ∐㎍∐짹∐∐∐∐㊄ upSubЀ⊏ЀЀ∐∐㎍㎍∐짹∐㎍∐뗄SubSubЀ⊏⊏퐿舠∐㎍∐⊏𓆝⊏㎍Sub넙Sub⊏㎍㎍㎍㎍∐∐㎍⊏∐⊏㎍짹SubⓍẀ⊏㎍㎍㎍㎍㎍㎍㎍⊏∐ⓙ⊏ⓙ∐∐ӝ⤙㎍㎍㎍㎍㎍㎍㎍㎍㎍ⓙⓙ∐𨨏ⓙ傛ⓙЀ⊏∐ϫꪑYesNoЀꪑ⍀⅗ upꪑ鐽⋉ꪑЀ鬈⊏넙蛲֩ the遹⊏蔟ᖃ of theϜ舠᭝Ẁ𨨏ଝNo⢺쫑쫑No穽𨨏Ꮁ∐舠堟⢺⋉ᶫ⢺粜钐뻑鏷🅤堟♼噚獓ߑ⅗∐𝟻ቍ遹立㎍ㅟ𝖿Ꮸፑ𝟻堟⡉毟∐⅗Subᅱ⪢∐㎍1㎍Ⅽ㎍≎ﻀ⪢ٮ꜄ⓙᅱ毟⊏㎍⊏ⓙ∐ⓙ∐ᔑ∐㎍Sub㎍SubⅭⓙ⸂∐ﻀ11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
```

I've created `self.tokenizer` in `Pi0.__init__` with:
```
from openpi.models.tokenizer import PaligemmaTokenizer
self.tokenizer = PaligemmaTokenizer(max_len=self.max_token_len)
```

Using a fully custom robot and didn't train my own tokenizer. Why do I get such a gibberish in `decoded_text`?

**Comment by ececyh** [comment]
I guess it needs autoregressive decoding as in https://github.com/Physical-Intelligence/openpi/issues/701. Just decoding prefix_out won't make it.
